File: src/main/java/leetcode/p038/Solution.java
package leetcode.p038;

class Solution {

    public static void main (String [] args){
        Solution sol = new Solution();
        for (int i = 1; i<10; i++)
            System.out.println(sol.countAndSay(i));
    }

    public String countAndSay(int n) {
        int init = 1;

        String res = "" + init;
        for (int i = 1; i< n; i++){
            res = count(res);
        }
        return res;
    }


    String count(String in){
        StringBuilder sb = new StringBuilder();

        Character cur = in.charAt(0);
        int count = 1;
        for (int i = 1; i < in.length(); i++){
            Character focus = in.charAt(i);
            if (focus.equals(cur))
                count ++;
            else {
                String app = count + "" + cur;
                sb.append(app);
                count = 1;
                cur = focus;
            }
        }
        String app = count + "" + cur;
        sb.append(app);
        return sb.toString();
    }
}

File: scala/src/leetcode/p038/Solution.scala
package leetcode.p038

object Solution {
  @main
  def main()={
    val x:BigInt = 3322251
    val rle = Rle.fromBigInt(x)
    println(s"$rle ${rle.toBigInt()} ${countAndSay(2)}")
  }

  def countAndSay(n: Int): String = {
    var cur: BigInt = 0
    for (i <- 1 to n){
       if  (i == 1)
        cur = 1
       else{
        cur = Rle.fromBigInt(cur).toBigInt()             
       }
    }
    cur.toString()       
  }
  

  object Rle {
    def fromBigInt(l: BigInt): Rle = {
      var remain: BigInt = l
      var numbers = List.empty[List[Int]]
      var count: Int = 1

      var prevdigit: Int = (remain % 10).toInt
      remain = remain / 10

      while ({
        val digit: Int = (remain % 10).toInt
        remain = remain / 10
        if (digit == prevdigit) {
          count = count + 1
        } else {
          numbers = List(count, prevdigit) +: numbers
          prevdigit = digit
          count = 1
        }
        remain > 0
      }) ()
      if (prevdigit != 0)
      numbers = List(count, prevdigit) +: numbers

      Rle(numbers)

    }
  }

  class Rle(val numbers: List[List[Int]]) // [#digits, digit]
  {
    override def toString(): String = {
        val sb: StringBuffer = StringBuffer()
        
        for (n <- numbers){
            sb.append(n(0).toString())
            sb.append(n(1).toString())
        }
        sb.toString()
    }
    def toBigInt(): BigInt = {
        var x: BigInt = 0
        for (n <- numbers){
            /*for (r <- 0 until n(0))
            {
                x = x * 10
                x = x + n(1) 
            }*/
            x = x * 10
            x = x + n(0)
            x = x * 10
            x = x + n(1)
        }
        x
    }
  }
}
